# Data Analysis with Pandas

###### Noemi Benci
Bachelor's Degree in Statistics.

###### Federico Pirona 
Bachelor's Degree in Statistics.

A little introduction to data if we want

## Loading data

In [1]:
import json
import pandas as pd

In [2]:
with open('./dati/dati-json/dpc-covid19-ita-province.json') as f:
    d1 = json.load(f)

In [3]:
with open('./dati/dati-json/dpc-covid19-ita-regioni.json') as f:
    d2 = json.load(f)

#### Visualize data as dictionary

In [15]:
d1[:2]

[{'data': '2020-02-24T18:00:00',
  'stato': 'ITA',
  'codice_regione': 13,
  'denominazione_regione': 'Abruzzo',
  'codice_provincia': 69,
  'denominazione_provincia': 'Chieti',
  'sigla_provincia': 'CH',
  'lat': 42.35103167,
  'long': 14.16754574,
  'totale_casi': 0,
  'note_it': '',
  'note_en': ''},
 {'data': '2020-02-24T18:00:00',
  'stato': 'ITA',
  'codice_regione': 13,
  'denominazione_regione': 'Abruzzo',
  'codice_provincia': 66,
  'denominazione_provincia': "L'Aquila",
  'sigla_provincia': 'AQ',
  'lat': 42.35122196,
  'long': 13.39843823,
  'totale_casi': 0,
  'note_it': '',
  'note_en': ''}]

In [16]:
d2[:2]

[{'data': '2020-02-24T18:00:00',
  'stato': 'ITA',
  'codice_regione': 13,
  'denominazione_regione': 'Abruzzo',
  'lat': 42.35122196,
  'long': 13.39843823,
  'ricoverati_con_sintomi': 0,
  'terapia_intensiva': 0,
  'totale_ospedalizzati': 0,
  'isolamento_domiciliare': 0,
  'totale_positivi': 0,
  'variazione_totale_positivi': 0,
  'nuovi_positivi': 0,
  'dimessi_guariti': 0,
  'deceduti': 0,
  'totale_casi': 0,
  'tamponi': 5,
  'casi_testati': None,
  'note_it': '',
  'note_en': ''},
 {'data': '2020-02-24T18:00:00',
  'stato': 'ITA',
  'codice_regione': 17,
  'denominazione_regione': 'Basilicata',
  'lat': 40.63947052,
  'long': 15.80514834,
  'ricoverati_con_sintomi': 0,
  'terapia_intensiva': 0,
  'totale_ospedalizzati': 0,
  'isolamento_domiciliare': 0,
  'totale_positivi': 0,
  'variazione_totale_positivi': 0,
  'nuovi_positivi': 0,
  'dimessi_guariti': 0,
  'deceduti': 0,
  'totale_casi': 0,
  'tamponi': 0,
  'casi_testati': None,
  'note_it': '',
  'note_en': ''}]

#### Visualize data as Pandas DataFrames

In [6]:
province = pd.DataFrame(d1)
province.sample(10)

data stato  codice_regione denominazione_regione  \
7013   2020-04-18T17:00:00   ITA              19               Sicilia   
8497   2020-04-30T17:00:00   ITA               3             Lombardia   
5398   2020-04-06T17:00:00   ITA               8        Emilia-Romagna   
7095   2020-04-19T17:00:00   ITA               3             Lombardia   
11919  2020-05-27T17:00:00   ITA              18              Calabria   
4775   2020-04-01T17:00:00   ITA              12                 Lazio   
4859   2020-04-01T17:00:00   ITA               5                Veneto   
2318   2020-03-13T17:00:00   ITA              18              Calabria   
306    2020-02-26T18:00:00   ITA               3             Lombardia   
6366   2020-04-13T17:00:00   ITA              19               Sicilia   

       codice_provincia               denominazione_provincia sigla_provincia  \
7013                994  In fase di definizione/aggiornamento                   
8497                 17                               Brescia              BS   
5398                 37                               Bologna              BO   
7095                 15                                Milano              MI   
11919               982  In fase di definizione/aggiornamento                   
4775                 57                                 Rieti              RI   
4859                 26                               Treviso              TV   
2318                102                         Vibo Valentia              VV   
306                  13                                  Como              CO   
6366                 87                               Catania              CT   

             lat       long  totale_casi note_it note_en  
7013    0.000000   0.000000            0                  
8497   45.539931  10.219103        12861                  
5398   44.494367  11.341721         2617                  
7095   45.466794   9.190347        15825                  
11919   0.000000   0.000000            2                  
4775   42.404884  12.862059          193                  
4859   45.667546  12.245074         1554                  
2318   38.676241  16.101574            5                  
306    45.809991   9.085160            0                  
6366   37.502878  15.087047          704

In [7]:
region = pd.DataFrame(d2)
region.sample(10)

data stato  codice_regione denominazione_regione  \
1185  2020-04-20T17:00:00   ITA               3             Lombardia   
451   2020-03-16T17:00:00   ITA              11                Marche   
191   2020-03-04T17:00:00   ITA              21          P.A. Bolzano   
1261  2020-04-24T17:00:00   ITA              17            Basilicata   
1850  2020-05-22T17:00:00   ITA              21          P.A. Bolzano   
1664  2020-05-13T17:00:00   ITA               8        Emilia-Romagna   
1614  2020-05-10T17:00:00   ITA              10                Umbria   
236   2020-03-06T17:00:00   ITA               8        Emilia-Romagna   
941   2020-04-08T17:00:00   ITA              22           P.A. Trento   
2312  2020-06-13T17:00:00   ITA              21          P.A. Bolzano   

            lat       long  ricoverati_con_sintomi  terapia_intensiva  \
1185  45.466794   9.190347                   10138                901   
451   43.616760  13.518875                     528                110   
191   46.499335  11.356624                       1                  0   
1261  40.639471  15.805148                      59                  7   
1850  46.499335  11.356624                      34                  5   
1664  44.494367  11.341721                    1165                123   
1614  43.106758  12.388247                      42                  6   
236   44.494367  11.341721                     397                 53   
941   46.068935  11.121231                     354                 77   
2312  46.499335  11.356624                       7                  2   

      totale_ospedalizzati  isolamento_domiciliare  totale_positivi  \
1185                 11039                   23548            34587   
451                    638                     547             1185   
191                      1                       0                1   
1261                    66                     163              229   
1850                    39                     185              224   
1664                  1288                    5214             6502   
1614                    48                      65              113   
236                    450                     366              816   
941                    431                    1509             1940   
2312                     9                      86               95   

      variazione_totale_positivi  nuovi_positivi  dimessi_guariti  deceduti  \
1185                          90             735            20008     12376   
451                           98             109                0        57   
191                            0               0                0         0   
1261                           0               4              107        24   
1850                         -26               3             2075       291   
1664                        -299              50            16572      3905   
1614                           2               4             1227        71   
236                          158             172               17        37   
941                           50             126              407       255   
2312                          -6               2             2224       291   

      totale_casi  tamponi  casi_testati note_it note_en  
1185        66971   270486      173333.0                  
451          1242     3225           NaN                  
191             1       20           NaN                  
1261          360     9026        9026.0                  
1850         2590    58188       26585.0                  
1664        26979   243883      155476.0                  
1614         1411    46923       32713.0                  
236           870     3136           NaN                  
941          2602    13258           NaN                  
2312         2610    75735       36393.0

#### First look to 'Province' Dataset
Discover info about the available variables and their nature.

In [8]:
province.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14720 entries, 0 to 14719
Data columns (total 12 columns):
data                       14720 non-null object
stato                      14720 non-null object
codice_regione             14720 non-null int64
denominazione_regione      14720 non-null object
codice_provincia           14720 non-null int64
denominazione_provincia    14720 non-null object
sigla_provincia            14720 non-null object
lat                        14720 non-null float64
long                       14720 non-null float64
totale_casi                14720 non-null int64
note_it                    14720 non-null object
note_en                    14720 non-null object
dtypes: float64(2), int64(3), object(7)
memory usage: 1.3+ MB


In [9]:
province.columns

Index(['data', 'stato', 'codice_regione', 'denominazione_regione',
       'codice_provincia', 'denominazione_provincia', 'sigla_provincia', 'lat',
       'long', 'totale_casi', 'note_it', 'note_en'],
      dtype='object')

In [10]:
province.describe(include='all')

data  stato  codice_regione denominazione_regione  \
count                 14720  14720     14720.00000                 14720   
unique                  115      1             NaN                    21   
top     2020-06-03T17:00:00    ITA             NaN             Lombardia   
freq                    128  14720             NaN                  1495   
mean                    NaN    NaN        10.68750                   NaN   
std                     NaN    NaN         6.12744                   NaN   
min                     NaN    NaN         1.00000                   NaN   
25%                     NaN    NaN         5.75000                   NaN   
50%                     NaN    NaN        10.00000                   NaN   
75%                     NaN    NaN        16.00000                   NaN   
max                     NaN    NaN        22.00000                   NaN   

        codice_provincia               denominazione_provincia  \
count       14720.000000                                 14720   
unique               NaN                                   108   
top                  NaN  In fase di definizione/aggiornamento   
freq                 NaN                                  2415   
mean          207.523438                                   NaN   
std           347.394828                                   NaN   
min             1.000000                                   NaN   
25%            32.750000                                   NaN   
50%            64.500000                                   NaN   
75%            96.250000                                   NaN   
max           999.000000                                   NaN   

       sigla_provincia           lat          long   totale_casi note_it  \
count            14720  14720.000000  14720.000000  14720.000000   14720   
unique             108           NaN           NaN           NaN       6   
top                              NaN           NaN           NaN           
freq              2415           NaN           NaN           NaN   14715   
mean               NaN     35.868843     10.112000   1159.016101     NaN   
std                NaN     16.072070      5.097765   2405.002333     NaN   
min                NaN      0.000000      0.000000      0.000000     NaN   
25%                NaN     38.848542      8.583304     34.000000     NaN   
50%                NaN     42.609815     11.188560    341.000000     NaN   
75%                NaN     44.947714     13.593962   1133.250000     NaN   
max                NaN     46.499335     18.171897  23966.000000     NaN   

       note_en  
count    14720  
unique       6  
top             
freq     14715  
mean       NaN  
std        NaN  
min        NaN  
25%        NaN  
50%        NaN  
75%        NaN  
max        NaN

The dataset of provinces is composed by 14 720 observations each composed by 12 variables. Each observation is a daily update of the count of cases of Coronavirus for a particular italian city and it provides all the information about the city (geographical location with longitude and latitude, which region does the city belong to) and the number of cases registred on a particular date. 

The dataset contains information of 115 different days for every region (21 unique regions) and province (107 provinces + 1 'In fase di aggiornamento') of Italy. 

On average have been registred 1159 cases per day, with a minimum of 0 and a maximum of about 24 000. 

##### First look to 'Regions' Dataset

In [11]:
region.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2415 entries, 0 to 2414
Data columns (total 20 columns):
data                          2415 non-null object
stato                         2415 non-null object
codice_regione                2415 non-null int64
denominazione_regione         2415 non-null object
lat                           2415 non-null float64
long                          2415 non-null float64
ricoverati_con_sintomi        2415 non-null int64
terapia_intensiva             2415 non-null int64
totale_ospedalizzati          2415 non-null int64
isolamento_domiciliare        2415 non-null int64
totale_positivi               2415 non-null int64
variazione_totale_positivi    2415 non-null int64
nuovi_positivi                2415 non-null int64
dimessi_guariti               2415 non-null int64
deceduti                      2415 non-null int64
totale_casi                   2415 non-null int64
tamponi                       2415 non-null int64
casi_testati                  1260 n

In [12]:
region.columns

Index(['data', 'stato', 'codice_regione', 'denominazione_regione', 'lat',
       'long', 'ricoverati_con_sintomi', 'terapia_intensiva',
       'totale_ospedalizzati', 'isolamento_domiciliare', 'totale_positivi',
       'variazione_totale_positivi', 'nuovi_positivi', 'dimessi_guariti',
       'deceduti', 'totale_casi', 'tamponi', 'casi_testati', 'note_it',
       'note_en'],
      dtype='object')

In [13]:
region.describe(include='all')

data stato  codice_regione denominazione_regione  \
count                  2415  2415     2415.000000                  2415   
unique                  115     1             NaN                    21   
top     2020-05-07T17:00:00   ITA             NaN        Emilia-Romagna   
freq                     21  2415             NaN                   115   
mean                    NaN   NaN       11.857143                   NaN   
std                     NaN   NaN        6.275098                   NaN   
min                     NaN   NaN        1.000000                   NaN   
25%                     NaN   NaN        7.000000                   NaN   
50%                     NaN   NaN       12.000000                   NaN   
75%                     NaN   NaN       17.000000                   NaN   
max                     NaN   NaN       22.000000                   NaN   

                lat         long  ricoverati_con_sintomi  terapia_intensiva  \
count   2415.000000  2415.000000             2415.000000        2415.000000   
unique          NaN          NaN                     NaN                NaN   
top             NaN          NaN                     NaN                NaN   
freq            NaN          NaN                     NaN                NaN   
mean      43.046293    12.225955              661.626501          75.839337   
std        2.489296     2.659498             1603.884235         169.204765   
min       38.115697     7.320149                0.000000           0.000000   
25%       41.125596    11.121231               28.000000           2.000000   
50%       43.616760    12.388247              127.000000          17.000000   
75%       45.434905    13.768136              526.000000          72.000000   
max       46.499335    16.867367            12077.000000        1381.000000   

        totale_ospedalizzati  isolamento_domiciliare  totale_positivi  \
count            2415.000000             2415.000000      2415.000000   
unique                   NaN                     NaN              NaN   
top                      NaN                     NaN              NaN   
freq                     NaN                     NaN              NaN   
mean              737.465839             2060.233540      2797.699379   
std              1766.756690             4164.175468      5654.196069   
min                 0.000000                0.000000         0.000000   
25%                34.000000              105.000000       151.000000   
50%               145.000000              621.000000       814.000000   
75%               604.000000             1965.000000      2678.500000   
max             13328.000000            30382.000000     37307.000000   

        variazione_totale_positivi  nuovi_positivi  dimessi_guariti  \
count                  2415.000000     2415.000000      2415.000000   
unique                         NaN             NaN              NaN   
top                            NaN             NaN              NaN   
freq                           NaN             NaN              NaN   
mean                      9.815321       98.477433      3309.270807   
std                     226.519785      242.944880      7714.324937   
min                   -5339.000000     -229.000000         0.000000   
25%                     -26.000000        2.000000        30.000000   
50%                       0.000000       16.000000       687.000000   
75%                      32.000000       80.000000      2472.500000   
max                    1950.000000     3251.000000     60850.000000   

            deceduti   totale_casi        tamponi   casi_testati note_it  \
count    2415.000000   2415.000000    2415.000000    1260.000000    2415   
unique           NaN           NaN            NaN            NaN      44   
top              NaN           NaN            NaN            NaN           
freq             NaN           NaN            NaN            NaN    2372   
mean      957.603727   7064.573913   85544.734990   93820.02698

The dataset of Regions contains 2415 rows and 20 variables each. Every observation is an update of some numbers about the spread of Coronavirus in Italy. 
For 115 days and for all italian regions we have:
- number of hospitalized with sympthoms (on average 661 per day, with minimum 0 and maximum of about 12 100);
- number of people in intensive care unit (on average 75 per day, with minimum 0 and maximum of about 1 400);
- total number of hospitalized (on average 737 per day, with minimum 0 and maximum 13 400);
- number of people house-isolated (on average 2 100 per day, with minimum 0 and maximum of about 30 400);
- number of tested positive  (on average 2 800 per day, with minimum 0 and maximum of about 37 300);
- variation of new tested positive (on average 9 per day, with minimum 0 and maximum of about 1 950);
- number of new tested positive (on average 98 per day, with minimum 0 and maximum of about 3 250);
- number discharged healed (on average 3 300 per day, with a minimum of 0 and a maximum of about 60 850);
- number of deaths (on average 950 per day, with a minimum of 0 and a maximum of about 16 500);
- number of total cases (on average 7 100 per day, with a minimum of 0 and a mamimum of about 92 300);
- number of swabs (on average 85 500 per day, with a minimum of 0 and a maximum of about 917 900);
- number of tested cases (on average 93 800 per day, with a minimum of 3 500 and a maximum of about 547 100).

This dataset can be integrated with the province dataset to integrate the information about the number of cases within the cities of a region.